daemon/gluster-blockd.c

```diff
@@ -43,6 +44,61 @@ static gbConfig *gbCfg;
 
 struct timeval TIMEOUT = {CLI_TIMEOUT_DEF, 0};  /* remote rpc call timeout, 5 mins */
 
+typedef struct globalData {
+  int    pfd;  /* pidfile file descriptor */
+  pid_t  chpid;
+} globalData;
+
+globalData ctx = {0,};
+
+
+void
+glusterBlockCleanGlobals(void)
+{
+  struct flock lock = {0, };
+
+
+  lock.l_type = F_UNLCK;
+  if (fcntl(ctx.pfd, F_SETLK, &lock) == -1) {
+    LOG("mgmt", GB_LOG_ERROR, "fcntl(UNLCK) on pidfile %s failed[%s]",
+        GB_LOCK_FILE, strerror(errno));
+  }
+
+  if (ctx.pfd) {
+    close(ctx.pfd);
+  }
+  glusterBlockDestroyConfig(gbCfg);
+}
+
+
+void
+onSigServerHandler(int signum)
+{
+  LOG("mgmt", GB_LOG_DEBUG,
+      "server process with (pid: %lu) received (signal: %s)",
+      getpid(), strsignal(signum));
+
+  svc_exit();
+
+  exit(EXIT_SUCCESS);
```

Comment:
There is no need to do the exit(EXIT_SUCCESS) here.

Reply:
@lxbsz

For some reason I see WEXITSTATUS() returning 1 if I don't enforce this here.
Which means an abnormal exit of child.

Can you check ?

Reply:
There is another problem:

[2019-04-15 11:54:52.545869] INFO: cli process pid: (18235) [at gluster-blockd.c+580 :<main>]
[2019-04-15 11:54:52.546274] INFO: server process pid: (18250) [at gluster-blockd.c+570 :<main>]
[2019-04-15 11:56:31.793613] INFO: Block Hosting Volfile Server Set to: localhost [at utils.c+276 :<fetchGlfsVolServerFromEnv>]
[2019-04-15 11:56:31.793715] CRIT: glfsLruCount now is 5 [at lru.c+42 :<glusterBlockSetLruCount>]
[2019-04-15 11:56:31.793729] CRIT: logLevel now is INFO [at utils.c+53 :<glusterBlockSetLogLevel>]
[2019-04-15 11:56:31.793920] INFO: Inotify is watching "/etc/sysconfig", wd: 1, mask: IN_MODIFY [at dyn-config.c+510 :<glusterBlockDynConfigStart>]
[2019-04-15 11:56:31.796159] INFO: Distro ID=fedora. Current kernel version: '4.18.16-300.fc29.x86_64'. [at gluster-blockd.c+421 :<gbMinKernelVersionCheck>]
[2019-04-15 11:56:35.381259] INFO: Block Hosting Volfile Server Set to: localhost [at utils.c+276 :<fetchGlfsVolServerFromEnv>]
[2019-04-15 11:56:35.381346] CRIT: glfsLruCount now is 5 [at lru.c+42 :<glusterBlockSetLruCount>]
[2019-04-15 11:56:35.381539] CRIT: logLevel now is INFO [at utils.c+53 :<glusterBlockSetLogLevel>]
[2019-04-15 11:56:35.382742] INFO: Inotify is watching "/etc/sysconfig", wd: 1, mask: IN_MODIFY [at dyn-config.c+510 :<glusterBlockDynConfigStart>]
[2019-04-15 11:56:35.384463] INFO: Distro ID=fedora. Current kernel version: '4.18.16-300.fc29.x86_64'. [at gluster-blockd.c+421 :<gbMinKernelVersionCheck>]
[2019-04-15 11:56:36.870342] INFO: capabilities fetched successfully [at gluster-blockd.c+550 :<main>]
[2019-04-15 11:56:36.871805] INFO: cli process pid: (19434) [at gluster-blockd.c+580 :<main>]
[2019-04-15 11:56:36.871881] INFO: server process pid: (19449) [at gluster-blockd.c+570 :<main>]
[2019-04-15 11:56:36.872043] ERROR: bind on port 24010 failed (Address already in use) [at gluster-blockd.c+263 :<glusterBlockServerProcess>]
[root@fedora2 gluster-block]# 


[root@fedora2 gluster-block]# systemctl status gluster-blockd tcmu-runner
● gluster-blockd.service - Gluster block storage utility
   Loaded: loaded (/usr/local/lib/systemd/system/gluster-blockd.service; enabled; vendor preset: disabled)
   Active: active (running) since Mon 2019-04-15 17:26:35 IST; 47s ago
 Main PID: 19434 (gluster-blockd)
    Tasks: 4 (limit: 2357)
   Memory: 2.7M
   CGroup: /system.slice/gluster-blockd.service
           ├─18250 /usr/local/sbin/gluster-blockd --glfs-lru-count 5 --log-level INFO
           └─19434 /usr/local/sbin/gluster-blockd --glfs-lru-count 5 --log-level INFO

I restart the gluster-blockd service, and the cli process is new, but the server process is still the old one.

Reply:
@lxbsz 
Did you notice the bind failure in the above log ?

It might have happen that your parent process exited before child in your previous run and child was attached to pid 1.

Can you run
\# pkill -9 gluster-blockd

then try again ?

Reply:
Please ignore this, this is should be the reason that I just kill -9 cli_process only and then restart it.

Reply:
> @lxbsz
> 
> For some reason I see WEXITSTATUS() returning 1 if I don't enforce this here.
> Which means an abnormal exit of child.
> 
> Can you check ?

From manual we can see that returning 1 is normal exit.

  ```
     WIFEXITED(status)
              returns true if the child terminated normally, that is, by calling exit(3) or _exit(2), or by returning from main().
```
That means the process is exit(EXIT_SUCCESS==0) or returned directly.

Reply:
I agree, but how can we consider child returning value 1 as success/normal ?

And, do you see anything wrong with exit(EXITSUCCESS) from child after svc_exit() ?

Reply:
Yeah, if we directly exit(EXIT_SUCCESS) here, then there maybe no chance to do one graceful exit for the child process, which will execute the code after svc_run() in server process.
Make sense ?

Reply:
@lxbsz 
Hmm, make sense to me, I got your concern here.
Please check the latest patch, which fixes it by moving the exit() part to glusterBlockServerProcess.